src/rgw/rgw_kmip_client_impl.cc

```diff
@@ -0,0 +1,729 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#include <boost/intrusive/list.hpp>
+#include <atomic>
+#include <mutex>
+#include <string.h>
+
+#include "include/compat.h"
+#include "common/errno.h"
+#include "rgw_common.h"
+#include "rgw_kmip_client.h"
+#include "rgw_kmip_client_impl.h"
+
+#include <openssl/err.h>
+#include <openssl/ssl.h>
+extern "C" {
+#include "kmip.h"
+#include "kmip_bio.h"
+#include "kmip_memset.h"
+};
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rgw
+
+static enum kmip_version protocol_version = KMIP_1_0;
+
+struct RGWKmipHandle {
+	int uses;
+	mono_time lastuse;
+	SSL_CTX *ctx;
+	SSL *ssl;
+	BIO *bio;
+	KMIP kmip_ctx[1];
+	TextString textstrings[2];
+	UsernamePasswordCredential upc[1];
+	Credential credential[1];
+	int need_to_free_kmip;
+	size_t buffer_blocks, buffer_block_size, buffer_total_size;
+	uint8 *encoding;
+
+	explicit RGWKmipHandle() :
+		uses(0), ctx(0), ssl(0), bio(0),
+		need_to_free_kmip(0),
+		encoding(0) {
+			memset(kmip_ctx, 0, sizeof kmip_ctx);
+			memset(textstrings, 0, sizeof textstrings);
+			memset(upc, 0, sizeof upc);
+			memset(credential, 0, sizeof credential);
+	};
+};
+
+struct RGWKmipWorker: public Thread {
+	RGWKMIPManagerImpl &m;
+	RGWKmipWorker(RGWKMIPManagerImpl& m) : m(m) {}
+	void *entry() override;
+	void signal() {
+		std::lock_guard l{m.lock};
+		m.cond.notify_all();
+	}
+};
+
+static void
+kmip_free_handle_stuff(RGWKmipHandle *kmip)
+{
+	if (kmip->encoding) {
+		kmip_free_buffer(kmip->kmip_ctx,
+			kmip->encoding,
+			kmip->buffer_total_size);
+		kmip_set_buffer(kmip->kmip_ctx, NULL, 0);
+	}
+	if (kmip->need_to_free_kmip)
+		kmip_destroy(kmip->kmip_ctx);
+	if (kmip->bio)
+		BIO_free_all(kmip->bio);
+	if (kmip->ctx)
+		SSL_CTX_free(kmip->ctx);
+}
+
+class RGWKmipHandleBuilder {
+private:
+	CephContext *cct;
+	const char *clientcert = 0;
+	const char *clientkey = 0;
+	const char *capath = 0;
+	const char *host = 0;
+	const char *portstring = 0;
+	const char *username = 0;
+	const char *password = 0;
+public:
+	RGWKmipHandleBuilder(CephContext *cct) : cct(cct) {};
+	RGWKmipHandleBuilder& set_clientcert(const std::string &v) {
+		const char *s = v.c_str();
+		if (*s) {
+			clientcert = s;
+		}
+		return *this;
+	}
+	RGWKmipHandleBuilder& set_clientkey(const std::string &v) {
+		const char *s = v.c_str();
+		if (*s) {
+			clientkey = s;
+		}
+		return *this;
+	}
+	RGWKmipHandleBuilder& set_capath(const std::string &v) {
+		const char *s = v.c_str();
+		if (*s) {
+			capath = s;
+		}
+		return *this;
+	}
+	RGWKmipHandleBuilder& set_host(const char *v) {
+		host = v;
+		return *this;
+	}
+	RGWKmipHandleBuilder& set_portstring(const char *v) {
+		portstring = v;
+		return *this;
+	}
+	RGWKmipHandleBuilder& set_username(const std::string &v) {
+		const char *s = v.c_str();
+		if (*s) {
```

Comment:
std::string::c_str() will not return null

Reply:
That's good, because if it did, "*s" would core dump.  What it's actually checking is that there is at least one non-nul character, ie, not the empty string.  The empty string is the default in the config logic and this code interprets that to mean "don't do password authentication", which is a required feature.